agent/xds/jwt_authn.go

```diff
@@ -84,67 +62,111 @@ func makeJWTAuthFilter(pCE map[string]*structs.JWTProviderConfigEntry, intention
+//
+// Note: since the rbac filter is in charge of making decisions of allow/denied, this
+// requirement uses `allow_missing_or_failed` to ensure it is always satisfied.
+func providerToJWTRequirement(provider *structs.JWTProviderConfigEntry) *envoy_http_jwt_authn_v3.JwtRequirement {
+	return &envoy_http_jwt_authn_v3.JwtRequirement{
+		RequiresType: &envoy_http_jwt_authn_v3.JwtRequirement_RequiresAny{
+			RequiresAny: &envoy_http_jwt_authn_v3.JwtRequirementOrList{
+				Requirements: []*envoy_http_jwt_authn_v3.JwtRequirement{
+					{
+						RequiresType: &envoy_http_jwt_authn_v3.JwtRequirement_ProviderName{
+							ProviderName: provider.Name,
+						},
+					},
+					// We use allowMissingOrFailed to allow rbac filter to do the validation
+					{
+						RequiresType: &envoy_http_jwt_authn_v3.JwtRequirement_AllowMissingOrFailed{
+							AllowMissingOrFailed: &emptypb.Empty{},
+						},
+					},
+				},
+			},
+		},
+	}
+}
 
+// collectJWTRequirements returns a list of all top level and permission level referenced providers.
+func collectJWTRequirements(i *structs.Intention) []*structs.IntentionJWTProvider {
+	var reqs []*structs.IntentionJWTProvider
+
+	// get permission level providers
 	reqs = append(reqs, getPermissionsProviders(i.Permissions)...)
 
+	if i.JWT != nil {
+		// get top level providers
+		reqs = append(reqs, i.JWT.Providers...)
+	}
+
 	return reqs
 }
 
-func getPermissionsProviders(p []*structs.IntentionPermission) []*jwtAuthnProvider {
-	var reqs []*jwtAuthnProvider
-	for k, perm := range p {
+func getPermissionsProviders(p []*structs.IntentionPermission) []*structs.IntentionJWTProvider {
+	var reqs []*structs.IntentionJWTProvider
+	for _, perm := range p {
 		if perm.JWT == nil {
 			continue
 		}
-		for _, prov := range perm.JWT.Providers {
-			reqs = append(reqs, &jwtAuthnProvider{Provider: prov, ComputedName: makeComputedProviderName(prov.Name, perm, k)})
-		}
+
+		reqs = append(reqs, perm.JWT.Providers...)
 	}
 
 	return reqs
 }
 
-// makeComputedProviderName is used to create names for unique provider per permission
-// This is to stop jwt claims cross validation across permissions/providers.
+// buildPayloadInMetadataKey is used to create a unique payload key per provider.
+// This is to ensure claims are validated/forwarded specifically under the right provider.
+// The forwarded payload is used with other data (eg. service identity) by the RBAC filter
+// to validate access to resource.
 //
-// eg. If Permission x is the 3rd permission and has a provider of original name okta
-// this function will return okta_3 as the computed provider name
-func makeComputedProviderName(name string, perm *structs.IntentionPermission, idx int) string {
-	if perm == nil {
-		return name
-	}
-	return fmt.Sprintf("%s_%d", name, idx)
+// eg. With a provider named okta will have a payload key of: jwt_payload_okta
+func buildPayloadInMetadataKey(providerName string) string {
+	return fmt.Sprintf("%s_%s", jwtMetadataKeyPrefix, providerName)
```

Comment:
nit: `xds` package is performance+memory sensitive because it's a hot path. Concatenating strings is more efficient here.

```suggestion
	return jwtMetadataKeyPrefix + "_" + providerName
```